sdks/wasm/binding_support.js

```diff
@@ -0,0 +1,363 @@
+
+			this.create_tcs = get_method ("CreateTaskSource");
+			this.set_tcs_result = get_method ("SetTaskSourceResult");
+			this.set_tcs_failure = get_method ("SetTaskSourceFailure");
+			this.tcs_get_task_and_bind = get_method ("GetTaskAndBind");
+
+			this.init = true;
+		},		
+
+		get_js_obj: function (js_handle) {
+			if (js_handle > 0)
+				return this.js_objects_table [js_handle - 1];
+			return null;
+		},
+		
+		//FIXME this is wastefull, we could remove the temp malloc by going the UTF16 route
+		//FIXME this is unsafe, cuz raw objects could be GC'd.
+		conv_string: function (mono_obj) {
+			if (mono_obj == 0)
+				return null;
+			var raw = this.mono_string_get_utf8 (mono_obj);
+			var res = Module.UTF8ToString (raw);
+			Module._free (raw);
+
+			return res;
+		},
+		
+		mono_array_to_js_array: function (mono_array) {
+			if (mono_array == 0)
+				return null;
+
+			var res = [];
+			var len = this.mono_array_length (mono_array);
+			for (var i = 0; i < len; ++i)
+				res.push (this.unbox_mono_obj (this.mono_array_get (mono_array, i)));
+
+			return res;
+		},
+
+		js_array_to_mono_array: function (js_array) {
+			var mono_array = this.mono_obj_array_new (js_array.length);
+			for (var i = 0; i < js_array.length; ++i) {
+				this.mono_obj_array_set (mono_array, i, this.js_to_mono_obj (js_array [i]));
+			}
+			return mono_array;
+		},
+
+		unbox_mono_obj: function (mono_obj) {
+			if (mono_obj == 0)
+				return undefined;
+			var type = this.mono_get_obj_type (mono_obj);
+			//See MARSHAL_TYPE_ defines in driver.c
+			switch (type) {
+			case 1: // int
+				return this.mono_unbox_int (mono_obj);
+			case 2: // float
+				return this.mono_unbox_float (mono_obj);
+			case 3: //string
+				return this.conv_string (mono_obj);
+			case 4: //vts
+				throw new Error ("no idea on how to unbox value types");
+			case 5: { // delegate
+				var obj = this.extract_js_obj (mono_obj);
+				return function () {
+					return BINDING.invoke_delegate (obj, arguments);
+				};
+			}
+			case 6: {// Task
+				var obj = this.extract_js_obj (mono_obj);
+				var cont_obj = null;
+				var promise = new Promise (function (resolve, reject) {
+					cont_obj = {
+						resolve: resolve,
+						reject: reject
+					};
+				});
+
+				this.call_method (this.setup_js_cont, null, "mo", [ mono_obj, cont_obj ]);
+				return promise;
+			}
+
+			case 7: // ref type
+				return this.extract_js_obj (mono_obj);
+			default:
+				throw new Error ("no idea on how to unbox object kind " + type);
+			}
+		},
+
+		create_task_completion_source: function () {
+			return this.call_method (this.create_tcs, null, "", []);
+		},
+
+		set_task_result: function (tcs, result) {
+			this.call_method (this.set_tcs_result, null, "oo", [ tcs, result ]);
+		},
+
+		set_task_failure: function (tcs, reason) {
+			this.call_method (this.set_tcs_failure, null, "os", [ tcs, reason.toString () ]);
+		},
+
+		js_to_mono_obj: function (js_obj) {
```

Comment:
This looks correct assuming the caller uses it correctly, but at some point I feel this function is going to be the function that throws when we have nulls/bugs propagate from elsewhere in the code. Some more error handling might make it easier to file bugs that throw here.

Reply:
This support code is only visible to the binding engine, users can't call it.
The only thing it's visible is call_method, cuz I exported it early, in the `BINDING__postset` part